connection.cc

```diff
@@ -75,25 +73,59 @@ void Connection::handle_request(const boost::system::error_code& error, size_t b
 			prefix = "Bad Request";
 		}
 		else {
-			// log what we're up to
-			request_summary_ = request->uri();
+			bool post_request_error = false;
+			
+			// read in the rest of a POST request
+			if (request->method() == "POST") {
+				std::string content_length_string = request->get_header("Content-Length");
+				if(content_length_string == "")
+				{
+					response.SetStatus(Response::bad_request);
+					response.SetBody("400 Bad Request");
+					post_request_error = true;
+				}
+				else
+				{
+					size_t content_length = std::stoi(content_length_string);
+
+					char buff[1000];
+					size_t body_size = request->body().size();
+					while (body_size < content_length) {
+						int nread;
+						try {
+							nread = socket_.read_some(boost::asio::buffer(buff, 1000));
+						}
+						catch (...){
+							// client closed the connection
+							// would close_socket() work?
+							return;
+						}
+						body_size = request->append_body(std::string(buff, nread)); // append the body and raw_request, update
+					}
+				}
+			}
 
-			// get the correct handler based on the header
-			const RequestHandler* handler = handlers_->Find(request->uri(), &prefix);
+			if(!post_request_error) {
+				// log what we're up to
+				request_summary_ = request->uri();
 
-			if (handler == nullptr) {
-				// TODO generalize, fit with the StaticHandler
-				NotFoundHandler not_found_handler;
-				not_found_handler.HandleRequest(*request, &response);
-				prefix = "Not Found";
-			}
-			else {
-				// have the handler generate a response
-				RequestHandler::Status status = handler->HandleRequest(*request, &response);
-				if (status != RequestHandler::OK)
-				{
-					response.SetStatus(Response::internal_server_error);
-					response.SetBody("500 Internal Server Error");
+				// get the correct handler based on the header
+				RequestHandler* handler = handlers_->Find(request->uri(), &prefix);
```

Comment:
Ahh what happened to all of the consts?
Do you know about mutable? That or you can store the state somewhere outside of the handler.

The constness helped guarantee thread safety. Race conditions on the shared state are now possible.